### PR TITLE
CPLAT-16481: Improve storage of pattern-matched mock HTTP request handlers

### PR DESCRIPTION
**Pull request by todbachman-wf**
## Motivation
We currently store pattern matching mock http request handlers in a `Map<Pattern, Map<String /* method */, PatternRequestHandler>>`. The key can be a RegExp. Two equivalent RegExp are not identical. Here's a quick example to demonstrate how this behaves:

```dart
void main() {
  final keyA = RegExp(r'https://fake.com/a/([bc]+)/d');
  final keyB = RegExp(r'https://fake.com/a/([bc]+)/d');
  final myMap = <Pattern, int>{};
  myMap[keyA] = 0;
  myMap[keyB] = 1;
  print(myMap.keys.length); // 2
  print(keyA == keyB); // false
}
```

This is usually OK. We run into an issue though if the consumer registers a pattern once, cancels it, and then registers it a second time with a newly constructed RegExp. In that case the previous pattern is still in the handler map, but it's value is an empty map of handlers. Here's a quick example:

```dart
final uriPattern = RegExp(r'https://fake.com/a/([bc]+)/d');
final mockHandler = MockTransports.http.whenPattern(uriPattern, _someHandler);

// The _patternRequestHandlers map will look something like this:
// {
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {
//.     '*': _someOtherHandler,
//.   }
// }

// later:
mockHandler.cancel();

// The _patternRequestHandlers map will look something like this:
// {
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {},
// }

// later yet:
final uriPattern = RegExp(r'https://fake.com/a/([bc]+)/d');
final mockHandler = MockTransports.http.whenPattern(uriPattern, _someOtherHandler);

// The _patternRequestHandlers map will look something like this:
// {
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {},
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {
//.     '*': _someOtherHandler,
//.   }
// }
```

Unfortunately, leaving the original RegExp key behind causes a problem in the _getMatchingHandler method. Specifically that method looks for the first key in the _patternRequestHandlers map that will match a URI. You can see in my example above that might be a key with a value that is an empty map. When this happens it fails to find an actual handler for the request and the request is left in a pending state.

Another situation in which this falls down is when registering handlers for multiple methods using the same regex pattern but not the same RegExp instances. Here's an example:

```dart
final uriPattern = RegExp( r'https://fake.com/a/([bc]+)/d');
final mockHandler = MockTransports.http.whenPattern(uriPattern, _getHandler, method: 'GET');

// The _patternRequestHandlers map will look something like this:
// {
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {
//.     'GET': _getHandler,
//.   }
// }

// later:
final uriPattern = RegExp( r'https://fake.com/a/([bc]+)/d');
final mockHandler = MockTransports.http.whenPattern(uriPattern, _postHandler, method: 'POST');

// The _patternRequestHandlers map will look something like this:
// {
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {
//.     'GET': _getHandler,
//.   },
//.   RegExp(r'https://fake.com/a/([bc]+)/d'): {
//.     'POST': _postHandler,
//.   }
// }
```

Unfortunately, because the same regex pattern appears in 2 keys in the map, when we look for the first pattern that matches a request it will maybe be the one with the GET handler but if the request is a POST we won't recognize there is a handler registered for it.

## Changes
- I've updated the cancel method on the MockHttpHandler that is returned to remove the key if the associated handler map is empty.
- I've introduced a `_Pattern` class that implements equality for 2 RegExp objects so they will be treated as the same key in our map of mock HTTP request handlers.

#### Release Notes
Remove URI pattern from mock http handlers map when the last handler is canceled.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/w_transport/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/w_transport/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by evanweible-wf — APPROVED** [review]

File: CHANGELOG.md
Comment (evanweible-wf) [inline CHANGELOG.md:1]:
```suggestion
## [4.0.1](https://github.com/Workiva/w_transport/compare/4.0.0...4.0.1)
```

**Review by evanweible-wf — APPROVED** [review]

**Comment by todbachman-wf** [comment]
I'm going to pull this into a couple repos to double check it doesn't cause any issues before merging.

**Comment by todbachman-wf** [comment]
QA +1. I pulled this into a couple internal repos and verified unit tests passed in both.

**Comment by todbachman-wf** [comment]
@Workiva/release-management-pp

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**
